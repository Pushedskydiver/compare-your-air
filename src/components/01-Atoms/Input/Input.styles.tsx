import styled from 'styled-components';
import { colors } from '../../../styles/colors.styles';
import { typography } from '../../../styles/typography.styles';
import spacing from '../../../styles/spacing.styles';

const InputStyles = styled.input`
  position: relative;
  width: 100%;
  padding-top: ${spacing.padding}px;
  padding-right: ${spacing.padding / 2}px;
  padding-bottom: ${spacing.padding}px;
  padding-left: ${spacing.padding * 4}px;
  border: 2px solid ${colors.neutral};
  border-radius: 8px;
  color: ${colors.dark};
  font-size: ${typography.body.size.regular};
  appearance: none;
  z-index: 2;
`;

export default InputStyles;
