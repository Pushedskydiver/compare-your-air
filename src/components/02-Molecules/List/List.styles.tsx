import styled from 'styled-components';
import { colors } from '../../../styles/colors.styles';
import { typography } from '../../../styles/typography.styles';
import spacing from '../../../styles/spacing.styles';

const ListStyles = styled.datalist`
  position: absolute;
  top: calc(100% - ${spacing.core}px);
  left: 0;
  width: 100%;
  max-height: 12em;
  margin: 0;
  padding-top: ${spacing.padding}px;
  padding-left: 0;
  border-bottom-right-radius: 6px;
  border-bottom-left-radius: 6px;
  background-color: ${colors.light};
  list-style: none;
  overflow-y: scroll;

  &:empty {
    padding-top: 0;
  }

  li {
    padding: ${spacing.padding}px;
    color: ${colors.dark};
    font-size: ${typography.body.size.regular};
    transition: background-color 0.3s ease-in-out;
    cursor: pointer;

    @media (prefers-reduced-motion: reduce) {
      transition: none;
    }
  }

  li[aria-selected="true"],
  li:hover {
    background-color: ${colors.pale};
  }
`;

export default ListStyles;
