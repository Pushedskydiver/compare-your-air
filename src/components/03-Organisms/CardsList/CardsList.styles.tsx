import styled from 'styled-components';
import spacing from '../../../styles/spacing.styles';

const CardsListStyles = styled.ul`
  display: grid;
  grid-gap: ${spacing.core * 1.5}px;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  margin-top: calc(12em + 20px);
  margin-bottom: ${spacing.margin * 4}px;
  padding-left: 0;
  list-style: none;
`;

export default CardsListStyles;
