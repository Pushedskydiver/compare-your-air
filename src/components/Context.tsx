import React, { createContext, useEffect, useState } from 'react';
import { fetchData } from './Utils/Utils';

type ProviderProps = {
  focusToInput?: boolean;
  searchResults?: Array<any>;
  searchTerm?: string;
  selectedLocations?: Array<any>;
  children: any
};

type LocationObject = {
  name: string
}

type LocationsArray = {
  results: Array<any>
}

export const Context = createContext({
  removeSelectedLocation: (location: string) => location,
  searchResults: [],
  searchTerm: '',
  selectLocation: (location: string) => location,
  selectedLocations: [],
  handleSearchTerm: (searchTerm: string) => searchTerm
});

export function ContextProvider(props: ProviderProps) {
  const locations = JSON.parse(localStorage.getItem('locations') || '[]');
  const [searchSate, setSearchState] = useState({
    searchTerm: '',
    searchResults: locations
  });

  const [resultsState, setResultsState] = useState({
    selectedLocations: []
  });

  function filterLocations(location: LocationObject, searchTerm: string) {
    return location.name.toLowerCase().includes(searchTerm.toLowerCase());
  }

  function handleSearchTerm(searchTerm: string) {
    const searchResults = locations.filter((location: LocationObject) => filterLocations(location, searchTerm));

    setSearchState({ ...searchSate, searchResults });

    return searchTerm;
  }

  function removeSelectedLocation(location: string) {
    const locations = resultsState.selectedLocations;
    const selectedLocations = locations.filter((item: any) => item.location !== location);

    setResultsState({ ...resultsState, selectedLocations });

    return location;
  }

  function removeDuplicateLocations(v: any, i: number, a: Array<any>) {
    return a.findIndex(t => (t.location === v.location)) === i;
  }

  function handleSelectedLocation({ results }: any) {
    const { selectedLocations } = resultsState;
    const mergedLocations: any = [...results, ...selectedLocations];
    const updatedLocations = mergedLocations.filter(removeDuplicateLocations);
    
    setResultsState({ ...resultsState, selectedLocations: updatedLocations });

    return selectedLocations;
  }

  function selectLocation(location: string) {
    const api = `https://api.openaq.org/v1/latest?city=${location}`;    
    
    fetchData(api)
      .then(handleSelectedLocation)
      .catch(error => console.error(error));

    return location;
  }

  useEffect(() => {
    const locationsData = localStorage.getItem('locations');

    function setLocations(locations: LocationsArray) {
      const searchResults = locations.results;      

      localStorage.setItem('locations', JSON.stringify(searchResults));
      setSearchState({ ...searchSate, searchResults });
    }

    function fetchCitiesData() {
      const api = 'https://api.openaq.org/v1/cities?country=GB';

      fetchData(api)
        .then(setLocations)
        .catch(error => console.error(error));
    }

    if (locationsData === null) {
      fetchCitiesData();
    }
  }, [searchSate]);

  const data = {
    ...searchSate,
    ...resultsState,
    handleSearchTerm,
    removeSelectedLocation,
    selectLocation,
  };

  return (
    <Context.Provider value={data}>
      {props.children}
    </Context.Provider>
  );
}

export default Context;