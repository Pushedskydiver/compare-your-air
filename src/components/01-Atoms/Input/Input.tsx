import React, { memo, useEffect, useContext, useRef, useState } from 'react';
import InputStyles from './Input.styles';
import Context from '../../Context';

const Input = () => {
  const { handleSearchTerm, handleShowResults, focusToInput, setInputFocus, showResults } = useContext(Context);
  const [value, setValue] = useState('Enter city name…');
  const inputRef = useRef<HTMLInputElement>(null);

  function handleChange(event: any) {
    const value: string = event.target.value;    

    if (!showResults) {
      handleShowResults(true);
    }

    setValue(value);
    handleSearchTerm(value);
  }

  function handleBlur() {
    if (value === '') {
      setValue('Enter city name…');
    }
  }

  function handleFocus() {
    if (value === 'Enter city name…') {
      setValue('');
      return;
    }

    if (value.length > 0) {
      handleSearchTerm(value);
    }
  }

  function handleClick() {
    if (!focusToInput) {
      setInputFocus(true);
    }

    handleShowResults(true);
  }

  function handleArrowDown(target: any) {
    if (!showResults) {
      handleShowResults(true);
    }

    setInputFocus(false);
    target.nextElementSibling.focus();
  }

  function handleKeyUpEvent(event: any) {
    if (event.key === 'ArrowDown') {
      handleArrowDown(event.target);
    }
  }

  function handleKeyDownEvent(event: any) {
    if (event.key === 'Tab') {
      setInputFocus(false);
    }
  }

  useEffect(() => {
    if (focusToInput && inputRef.current !== null) {
      inputRef.current.focus();
    }
  }, [focusToInput, inputRef]);

  return (
    <InputStyles
      type="text"
      id="city"
      ref={inputRef}
      role="combobox"
      value={value}
      onChange={handleChange}
      onBlur={handleBlur}
      onFocus={handleFocus}
      onClick={handleClick}
      onKeyUp={handleKeyUpEvent}
      onKeyDown={handleKeyDownEvent}
      aria-autocomplete="list"
      aria-owns="results"
      autoComplete="off"
    />
  );
}

export default memo(Input);
