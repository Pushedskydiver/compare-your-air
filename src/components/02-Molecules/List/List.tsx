import React, { useContext } from 'react';
import ListStyles from './List.styles';
import Context from '../../Context';

type SearchResultsProps = {
  name: string
};

const List = () => {
  const { searchResults } = useContext(Context);

  function renderCity(location: SearchResultsProps, index: number) {    
    return (
      <option
        key={index}
        value={location.name}
      />
    );
  }

  return (
    <ListStyles
      data-testid="location-results"
      id="results"
    >
      {searchResults.map(renderCity)}
    </ListStyles>
  )
}

export default List;
