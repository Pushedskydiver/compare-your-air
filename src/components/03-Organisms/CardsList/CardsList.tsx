import React, { useContext } from 'react';
import CardsListStyles from './CardsList.styles';
import Card from '../../02-Molecules/Card/Card';
import Context from '../../Context';

type PropsForLocation = {
  location: string;
  city: string;
  measurements: Array<any>;
};

const CardsList = () => {
  const { selectedLocations } = useContext(Context);

  function renderLocation(data: PropsForLocation, index: number) {
    const { location, city, measurements } = data;

    if (selectedLocations.length > 0) {
      return (
        <Card
          key={index}
          location={location}
          city={city}
          measurements={measurements}
        />
      )
    }

    return '';
    
  }

  return (
    <CardsListStyles data-testid="locations-list">
      {selectedLocations.map(renderLocation)}
    </CardsListStyles>
  );
}

export default CardsList;
