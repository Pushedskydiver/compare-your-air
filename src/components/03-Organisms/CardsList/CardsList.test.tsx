import React from 'react';
import { cleanup, render, waitForDomChange } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from '../../01-Atoms/Input/Input';
import List from '../../02-Molecules/List/List';
import CardsList from './CardsList';
import { ContextProvider } from '../../Context';

const data = [
  {
    country: 'GB',
    name: 'Aberdeen',
    city: 'Aberdeen',
    count: 176315,
    locations: 3
  },
  {
    country: 'GB',
    name: 'Armagh',
    city: 'Armagh',
    count: 54863,
    locations: 1
  },
  {
    country: 'GB',
    name: 'Aston Hill',
    city: 'Aston Hill',
    count: 57542,
    locations: 1
  }
];

beforeAll(() => {
  localStorage.setItem('locations', JSON.stringify(data));
})

afterEach(() => {
  cleanup();
});

describe('<CardsList />', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <CardsList />
    );
    expect(container).toMatchSnapshot();
  });

  it('should output cities in Aberdeen to locations list', async () => {
    const { container, getByTestId } = render(
      <ContextProvider>
        <Input />
        <List />
        <CardsList />
      </ContextProvider>
    );

    const input = getByTestId('location-input');
    const results = getByTestId('location-results');
    const locations = getByTestId('locations-list');

    userEvent.click(input);
    await userEvent.type(input, 'Aberdeen');
    userEvent.click(results.children[0]);
    await waitForDomChange({ container });

    expect(locations.children.length).toBeGreaterThan(0);
  });
})