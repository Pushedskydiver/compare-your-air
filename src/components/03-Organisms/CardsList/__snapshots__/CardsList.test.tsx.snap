// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardsList /> should match snapshot 1`] = `
<div>
  <ul
    class="sc-AxiKw fteBzX"
    data-testid="locations-list"
  />
</div>
`;
